docs/DesignDocs/system_ocall_opt_in.md

```diff
@@ -0,0 +1,120 @@
+
+Because ocalls have potential security impact, and enclave developer should be able to opt
+out of these ocalls being made and should have visibility into what ocalls might be called
+by OE SDK.
+
+## EDL installation
+
+In order to support ocall opt-in, system EDL files will now be installed as part of the SDK at
+`${install_path}/include/openenclave/edl/`. This allows them to be imported into user edl by the
+enclave developer.
+
+## Developer opt-in
+
+In order to use any system ocall, it must be imported into the enclave's edl file. If it is not
+included, but the ocall is still called by the enclave application or some part of the SDK,
+enclave linking will fail.
+
+Users can import a system ocall the same way any other edl file would be imported. For example,
+if an enclave with edl file `sample.edl` wants to use the function `oe_log_ocall()` they can do
+one of the following:
+
+*Note: These examples assume that `${install_path}/include` is passed to edger8r as an edl search path*
+
+1. Import just the required ocall
+
+```
+// sample.edl
+enclave
+{
+    from "openenclave/edl/tee.edl" import oe_log_ocall;
+
+    trusted
+    {
+        public void my_sample_ecall();
+    };
+
+    untrusted
+    {
+        public void my_sample_ocall();
+    };
+}
+```
+
+2. Import all ocalls from `tee.edl`
+
+```
+// sample.edl
+enclave
+{
+    from "openenclave/edl/tee.edl" import *;
+
+    trusted
+    {
+        public void my_sample_ecall();
+    };
+
+    untrusted
+    {
+        public void my_sample_ocall();
+    };
+}
+```
+
+In example (1), linking would fail if the enclave made a call to `oe_realloc_ocall()`
+(also part of tee.edl) unless the developer modified the import line to
+`from "openenclave/edl/tee.edl" import oe_log_ocall, oe_realloc_ocall;`. In this scnario,
+example (2) will work without modification.
+
+## EDL local structures
+
+Under this design, system edl files will prefer using local structures; that is, structures
+that are defined in edl. This is distinct from foreign structures, which are defined in
+C/C++ header files and included in the edl file.
+
+In the case that a foreign structure must be used, system edl files may only include headers
+that are installed with the SDK.
+
+Local EDL structures provide additional security benefits detailed [here](
+https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/full_edl_serialization.md#motivation)
+so foreign structures should be avoided whenever possible.
+
+## Developer opt-out
+
+Because system ocalls are generally called by the SDK implementation, developers
+may not have direct control over which ocalls will be called by their application.
+A developer can refuse to import a function in EDL which will cause a linker error
+if it is used, but they should also be able to take action to resolve this failure.
+
+To this end, any OE SDK feature which makes an ocall on behalf of an enclave
```

Comment:
I'm not sure I understand what the opt-out option entails here. It seems like disabling a feature at compile time would be directly equivalent to refactoring the feature so that it could be explicitly opted into. Do you have a proposed example for such a mechanism to illustrate the difference you have in mind?

It seems like instead of introducing the opt-in/opt-out dichotomy, it would be more informative to propose how we would handle existing exceptions to this model, see other comments below.

Reply:
Reworked to hopefully be more clear